feat(kms-connector): retrieve s3 url in extra_data

eudelins-zama opened the pull request:
Ref https://github.com/zama-ai/fhevm-internal/issues/404

eudelins-zama committed:
feat(kms-connector): retrieve s3 url in extra_data

chore(kms-connector): update mock contract address

chore(kms-connector): remove gw conn from kms-worker

chore(kms-connector): update extra_data encoding

fix(kms-connector): fix kms-worker health tests

chore(kms-connector): adapt storage urls retrieval

chore(kms-connector): s3 urls edge case handling

melanciani commented on kms-connector/crates/kms-worker/src/core/event_processor/s3.rs:
```diff
@@ -406,6 +216,17 @@ pub fn compute_digest(ct: &[u8]) -> Vec<u8> {
     result
 }
 
+pub fn extract_fallback_url(s3_config: &S3Config) -> String {
```
are we sure we want a fallback url ? what's the purpose of it exactly ?

eudelins-zama replied:
It has been useful in the past when zws had issue with S3 setup, it offered a way to use the S3 bucket we wanted as a fallback to debug the rest of the stack if needed

melanciani reviewed (approved):
thanks !

melanciani commented on kms-connector/crates/kms-worker/src/core/event_processor/decryption.rs:
```diff
@@ -118,14 +114,14 @@ where
+        extra_data: &[u8],
     ) -> anyhow::Result<Vec<TypedCiphertext>> {
-        // Retrieve ciphertext materials from S3
+        let s3_url_matrix = decode_s3_url_matrix(extra_data)?;
```
I updated https://github.com/zama-ai/fhevm-internal/issues/404#issuecomment-3323575775 : the urls are no longer in extraData and instead are a new field to the event

melanciani reviewed (changes requested):
small change: storage urls are no longer in extraData and can be found as a new field in the request event

eudelins-zama commented:
Should be good now @RomanBredehoft (cc @dartdart26)

dartdart26 committed:
feat(coprocessor): add copro context ID for input proofs

Add support for a separate uin256 context ID in the event listened to
by the gw-listener and using it to sign a `CiphertextVerification` in
transaction-sender.

Related to https://github.com/zama-ai/fhevm/pull/211

melanciani commented on kms-connector/crates/kms-worker/src/core/event_processor/s3.rs:
```diff
@@ -37,208 +27,42 @@ pub struct S3Service<P: Provider> {
 
-        for sns_material in sns_materials {
-            // Get S3 URL and retrieve ciphertext
+        for (sns_material, mut s3_urls) in sns_materials.into_iter().zip(s3_urls_matrix) {
```
shouldn't you maybe check that both `sns_materials` and `s3_urls_matrix` have the sane length with a proper error ? just in case ?

eudelins-zama replied:
Good catch! I did in aaea886bfd39fc358a3106943b151f7e9f3c1cc6 

Not raising errors immediately, will try to use fallback config first FYI

melanciani replied:
alright, but still not sure I'm a fan of the fallback url : how do you chose the coprocessor storage to use for the fallback ? 

this means we somewhat have to define/harcode a "more trusted" copro for its storage url. But this feels a bit wrong no ? What if the chosen coprocessor becomes malicious, or its storage URL changes ? It looks like the gateway is no longer the single point of truth as these cases will require additional manual update in each KMS connector

What I'm trying to say is that I feel like if there is an issue with the URLs sent by the gateway, it's most likely an issue in the protocol that should raise an error and that should be fixed in the gateway itself. I don't think the connector should have a patch for these cases 

wdyt ?

eudelins-zama replied:
For now, this was just used by infra as a debugging solution on devnet/testnet, not meant to be used once the env is working well.

But we could open an issue to completely remove it if it is OK on infra side, I just don't think this PR is the good place for that, so I tried to make it coherent with current implementation

melanciani replied:
yes I see, well yes I'll let you create an issue on that because I indeed we shouldn't keep this code, but yes agree maybe not in this PR !

melanciani reviewed (approved):
looks good to me !

isaacdecoded committed:
chore(common): sync with main

isaacdecoded committed:
chore(common): sync with main

melanciani committed:
chore(common): sync with main

eudelins-zama commented:
Closing, no longer wanted
